### PR TITLE
API 4.7

### LINKED ISSUE #1856: API 4.7
Pure API updates:

- [x] Add class `Dice`
- [x] Add `Bot.send_dice`
- [x] Add `Message.dice`
- [x] Add `Message.reply_dice`
- [x] Add class `BotCommand`
- [x] Add `Bot.get_my_commands`
- [x] Add `Bot.set_my_commands`
- [x] Add `Bot.commands` property
- [x] Add parameter `tgs_sticker` to `Bot.create_new_sticker_set`
- [x] Add parameter `tgs_sticker` to `Bot.add_sticker_to_set`
- [x] Add `StickerSet.thumb`
- [x] Add `Bot.set_sticker_set_thumb`
- [ ] Add warning about changed parameter order for `Bot.create_new_sticker_set` and `Bot.add_sticker_to_set` to release notes

My suggestions for according new functionality:

- [ ] Add `Filters.dice` to filter messages containing a dice
- [ ] Allow `Filters.dice(1)` and `Filters.dice([1, 5])` to filters for dices with specific values
- [ ] Somehow allow to treat a dice message as text message with the corresponding emoji via `Filters.text` and `Message.effective_message`?
- [x] ~Maybe add a `description` attribute to `CommandHandler` and a method `Dispatcher.set_bot_commands` that automatically sets the bot commands by retrieving them from the registered handlers?~

### PR DESCRIPTION
**Pull request by Bibo-Joshi**
Closes #1856 

Pure API updates:

- [x] Add class `Dice`
- [x] Add `Bot.send_dice`
- [x] Add `Message.dice`
- [x] Add `Message.reply_dice`
- [x] Add class `BotCommand`
- [x] Add `Bot.get_my_commands`
- [x] Add `Bot.set_my_commands`
- [x] Add `Bot.commands` property
- [x] Add parameter `tgs_sticker` to `Bot.create_new_sticker_set`
- [x] Add parameter `tgs_sticker` to `Bot.add_sticker_to_set`
- [x] Add `StickerSet.thumb`
- [x] Add `Bot.set_sticker_set_thumb`
- [x] Tests
- [ ] Add warning about changed parameter order for `Bot.create_new_sticker_set` and `Bot.add_sticker_to_set` to release notes

New functionality:

- [x] Add `Filters.dice` to filter messages containing a dice
- [x] Allow `Filters.dice(1)` and `Filters.dice([1, 5])` to filters for dices with specific values
- [x] Somehow allow to treat a dice message as text message with the corresponding emoji via `Filters.text` and `Message.effective_message`?
- [x] Tests

Problems I encountered:
- I didn't have the means to nor did I feel like creating a custom animated sticker for testing purposes. So I just downloaded one vie TG desktop. Copyright concerns?
- `Bot.create_new_sticker_set` was not tested before, I guess because we can't delete packs via the API. So I don't test with the new parameter, either.
- ~`Bot.upload_sticker_file` seems to accept TGS files as `png_sticker`. I'm jusing it it `test_bot_methods_1_tgs` right now. Also pinged @botsupport about it.~ The resulting `file_id` is not valid
- While improving coverage I noticed, that the `mask_position` for `add_sticker_to_set` was untested. If I see correctly, request posts differently when posting files, which is the case for stickers, and thus the `mask_position` has to be `to_json`-ed manually. Added that and the corresponding test.

### CODE REVIEW

File: telegram/bot.py [hunk 1]
```diff
@@ -3308,15 +3319,23 @@ def upload_sticker_file(self, user_id, png_sticker, timeout=20, **kwargs):
         return File.de_json(result, self)
 
     @log
-    def create_new_sticker_set(self, user_id, name, title, png_sticker, emojis,
-                               contains_masks=None, mask_position=None, timeout=20, **kwargs):
+    def create_new_sticker_set(self, user_id, name, title, emojis, png_sticker=None,
+                               contains_masks=None, mask_position=None, tgs_sticker=None,
```
Comment (tsnoam) [inline telegram/bot.py]: Add new args last, helps with backward comparability.

Comment (tsnoam) [inline telegram/bot.py]: Add new arguments last, for backward compatibility

File: telegram/bot.py [hunk 2]
```diff
@@ -3368,23 +3394,35 @@ def create_new_sticker_set(self, user_id, name, title, png_sticker, emojis,
         return result
 
     @log
-    def add_sticker_to_set(self, user_id, name, png_sticker, emojis, mask_position=None,
-                           timeout=20, **kwargs):
-        """Use this method to add a new sticker to a set created by the bot.
+    def add_sticker_to_set(self, user_id, name, emojis, png_sticker=None, mask_position=None,
+                           tgs_sticker=None, timeout=20, **kwargs):
```
Comment (tsnoam) [inline telegram/bot.py]: Add new args last

File: telegram/bot.py [hunk 3]
```diff
@@ -3470,6 +3512,48 @@ def delete_sticker_from_set(self, sticker, timeout=None, **kwargs):
...
+            **kwargs (:obj:`dict`): Arbitrary keyword arguments.
+
+        Returns:
+            :obj:`bool`: On success, ``True`` is returned.
+
+        Raises:
+            :class:`telegram.TelegramError`
+
+        """
+        url = '{0}/setStickerSetThumb'.format(self.base_url)
```
Comment (tsnoam) [inline telegram/bot.py]: Just '{}' we don't need to support ancient python versions

File: telegram/bot.py [hunk 4]
```diff
@@ -3768,9 +3950,17 @@ def __reduce__(self):
...
     sendPoll = send_poll
     """Alias for :attr:`send_poll`"""
     stopPoll = stop_poll
     """Alias for :attr:`stop_poll`"""
+    sendDice = send_dice
+    """Alias for :attr:`send_dice`"""
+    getMyCommands = get_my_commands
+    """Alias for :attr:`get_my_commands`"""
+    setMyCommands = set_my_commands
+    """Alias for :attr:`set_my_commands`"""
```
Comment (tsnoam) [inline telegram/bot.py]: For some reason I remembered that we deprecated the camelCase aliases 🤔

Oh well...

Comment (Bibo-Joshi) [reply telegram/bot.py]: Not that I know of. Maybe for v13? We only need to adjust `test_official`, but that should be doable

Comment (tsnoam) [reply telegram/bot.py:3981]: Doesn't really matter except I'm hallucinating

File: telegram/ext/filters.py [hunk 5]
```diff
@@ -957,6 +957,95 @@ def filter(self, message):
...
+    Examples:
+        To allow any dice message, simply use
+        ``MessageHandler(Filters.dice, callback_method)``.
+        To allow only dice with value 6, use
+        ``MessageHandler(Filters.dice(6), callback_method)``.
+        To allow only dice with value 5 `or` 6, use
+        ``MessageHandler(Filters.dice([5, 6]), callback_method)``.
+
+    Args:
+        update (:obj:`int` | Iterable[:obj:`int`], optional): Which values to allow. If not
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:1033]: Okay. I pretty much copy-pasted from `Filters.text`, where this was introduced in #1631. Should I change the naming there, too (in another PR)?

File: telegram/ext/filters.py [hunk 6]
```diff
@@ -957,6 +957,95 @@ def filter(self, message):
...
+        class _DiceIterable(BaseFilter):
+
+            def __init__(self, iterable):
+                if isinstance(iterable, int):
+                    self.iterable = [iterable]
+                else:
+                    self.iterable = iterable
+                self.name = 'Filters.dice({})'.format(iterable)
+
+            def filter(self, message):
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:972]: Personally, I'd like to have pre-commit and CI set up before … See #1373

File: telegram/ext/filters.py [hunk 7]
```diff
@@ -957,6 +957,95 @@ def filter(self, message):
...
+        To allow only dice with value 5 `or` 6, use
+        ``MessageHandler(Filters.dice([5, 6]), callback_method)``.
+
+    Args:
+        update (:obj:`int` | Iterable[:obj:`int`], optional): Which values to allow. If not
+            specified, will allow any dice message.
+
+    Attributes:
+        text: Filters for dice messages, but allows to treat them as text messages, i.e. if the
+            dice messages text is empty, it will be set to the dice emoji ``🎲``. Like
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:1038]: Okay. Was just an idea. I'll still keep `Filters.dice.text` to treat dice messages as text messages while filtering and add a note that they won't actually have a text, if that's okay

File: telegram/ext/filters.py [hunk 8]
```diff
@@ -957,6 +957,95 @@ def filter(self, message):
...
+                else:
+                    return self._DiceTextIterable(update)
+
+        text = _DiceText()
+
+        def __call__(self, update):
+            if isinstance(update, Update):
+                return self.filter(update.effective_message)
+            else:
+                return self._DiceIterable(update)
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:1015]: See #1631 for an explanation

File: telegram/dice.py [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+# You should have received a copy of the GNU Lesser Public License
+# along with this program.  If not, see [http://www.gnu.org/licenses/].
+"""This module contains an object that represents a Telegram Dice."""
+from telegram import TelegramObject
+
+
+class Dice(TelegramObject):
+    """
+    This object represents a dice with random value from 1 to 6. (Yes, Telegram is aware of the
+    proper singular of die. They don't like it.)
```
Comment (tsnoam) [inline telegram/dice.py:27]: that's good and valid English but I actually had to open a dictionary. maybe we can rephrase that comment?

File: telegram/ext/filters.py [hunk 10]
```diff
@@ -272,6 +272,10 @@ def filter(self, message):
             ...
             MessageHandler(Filters.text(buttons), callback_method)
 
+    Note:
+        Dice messages don't have text. If you want to treat them as text messages anyway, use
+        ``Filters.text | Filters.dice``.
```
Comment (tsnoam) [inline telegram/ext/filters.py:277]: Is that accurate? How does that work?

File: telegram/ext/filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tsnoam) [reply telegram/ext/filters.py]: Looking at the code, it's not the only filter. Please open an issue for all of them.

File: telegram/ext/filters.py [hunk 11]
```diff
@@ -957,6 +957,95 @@ def filter(self, message):
...
+        name = 'Filters.dice'
+
+        class _DiceIterable(BaseFilter):
+
+            def __init__(self, iterable):
+                if isinstance(iterable, int):
+                    self.iterable = [iterable]
+                else:
+                    self.iterable = iterable
+                self.name = 'Filters.dice({})'.format(iterable)
```
[the remark this answers is not in the document]
Comment (tsnoam) [reply telegram/ext/filters.py]: I'm willing to make that sacrifice.

File: telegram/ext/filters.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/ext/filters.py]: #1881

File: telegram/dice.py (hunk 9 above)
Comment (Bibo-Joshi) [reply telegram/dice.py:27]: Rephrased to "Yes, Telegram is aware of the proper singular of dice, namely "die". They don't like it."

File: telegram/ext/filters.py (hunk 10 above)
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:277]: Yeah, that's rather a remnant of my previous attempt. Rephrased.

**Comment by Bibo-Joshi** [comment]
> we're getting close. a few minor details and we're set to go.

About the not-tested lines:

one chunk is `Bot.create_new_sticker_set`, which wasn't tested before (see PR desc)
the other is reply_markup in which also wasn't tested before (introduced in #1418 ). I can try and test it

File: telegram/ext/filters.py (hunk 10 above)
Comment (tsnoam) [reply telegram/ext/filters.py:277]: If you want to __filter__ either ...

File: telegram/ext/filters.py [hunk 12]
```diff
@@ -999,7 +999,7 @@ def filter(self, message):
             specified, will allow any dice message.
 
         Note:
-            Dice messages don't have text. If you want to treat them as text messages anyway, use
+            Dice messages don't have text. If you want to either text or dice messages, use
```
Comment (tsnoam) [inline telegram/ext/filters.py:1002]: Same comment as above

File: telegram/botcommand.py [hunk 13]
```diff
@@ -0,0 +1,47 @@
...
+
+
+class BotCommand(TelegramObject):
+    """
+    This object represents a bot command.
+
+    Attributes:
+        command (:obj:`str`): Text of the command, 1-32 characters. Can contain only lowercase
+            English letters, digits and underscores.
+        description (:obj:`str`): Description of the command, 3-256 characters.
```
Comment (Poolitzer) [inline telegram/botcommand.py:31]: Last discussion I had with the other maintainers about Attributes and Args was that Attributes are supposed to be shorter, so uhm, I suggest to make them shorter.

```suggestion
        command (:obj:`str`): Text of the command.
        description (:obj:`str`): Description of the command.
```

File: telegram/dice.py [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+from telegram import TelegramObject
+
+
+class Dice(TelegramObject):
+    """
+    This object represents a dice with random value from 1 to 6. (The singular form of "dice" is
+    "die". However, PTB mimics the Telegram API, which uses the term "dice".)
+
+    Attributes:
+        value (:obj:`int`): Value of the dice, 1-6.
```
Comment (Poolitzer) [inline telegram/dice.py:30]: Attributes are supposed to be shorter

```suggestion
    Attributes:
        value (:obj:`int`): Value of the dice.
```

File: telegram/ext/filters.py [hunk 15]
```diff
@@ -957,6 +961,48 @@ def filter(self, message):
...
+                return self.filter(update.effective_message)
+            else:
+                return self._DiceValues(update)
+
+        def filter(self, message):
+            return bool(message.dice)
+
+    dice = _Dice()
+    """Dice Messages. If an integer or a list of integers is passed, it filters messages to only
+    allow those whose dice value is appearing in the given list.
```
Comment (Poolitzer) [inline telegram/ext/filters.py:987]:
```suggestion
    allow those with dice where the value is appearing in the given list.
```

File: telegram/message.py [hunk 16]
```diff
@@ -106,6 +106,7 @@ class Message(TelegramObject):
         passport_data (:class:`telegram.PassportData`): Optional. Telegram Passport data.
         poll (:class:`telegram.Poll`): Optional. Message is a native poll,
             information about the poll.
+        dice (:class:`telegram.Dice`): Optional. Message is a dice with random value from 1 to 6.
```
Comment (Poolitzer) [inline telegram/message.py:109]: Attributes are supposed to be shorter

```suggestion
        dice (:class:`telegram.Dice`): Optional. Message is a dice.
```

File: telegram/message.py [hunk 17]
```diff
@@ -808,6 +813,23 @@ def reply_poll(self, *args, **kwargs):
         self._quote(kwargs)
         return self.bot.send_poll(self.chat_id, *args, **kwargs)
 
+    def reply_dice(self, *args, **kwargs):
+        """Shortcut for::
+
+            bot.send_dice(update.message.chat_id, *args, **kwargs)
+
+        Keyword Args:
+            quote (:obj:`bool`, optional): If set to ``True``, the poll is sent as an actual reply
```
Comment (Poolitzer) [inline telegram/message.py:822]: I know where you copied this from

```suggestion
            quote (:obj:`bool`, optional): If set to ``True``, the dice is sent as an actual reply
```

File: telegram/bot.py [hunk 18]
```diff
@@ -3308,15 +3321,23 @@ def upload_sticker_file(self, user_id, png_sticker, timeout=20, **kwargs):
...
         Note:
-            The png_sticker argument can be either a file_id, an URL or a file from disk
-            ``open(filename, 'rb')``
+            The png_sticker and tgs_sticker argument can be either a file_id, an URL or a file from
+            disk ``open(filename, 'rb')``
+
+        Warning:
+            As of API 4.7 ``png_sticker`` is an optional argument and therefore the order of the
+            arguments had to be changed. Use keyword arguments to make sure that the arguments are
+            passed correctly.
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/bot.py:3340]: Changed

File: telegram/bot.py [hunk 19]
```diff
@@ -3354,37 +3379,55 @@ def create_new_sticker_set(self, user_id, name, title, png_sticker, emojis,
...
```
[the remark this answers is not in the document]
Comment (Bibo-Joshi) [reply telegram/bot.py:3416]: changed

File: telegram/botcommand.py (hunk 13 above)
Comment (Bibo-Joshi) [reply telegram/botcommand.py:31]: you're right, we're mostly not stating the restrictions at the attributes. changed

File: telegram/dice.py (hunk 14 above)
Comment (Bibo-Joshi) [reply telegram/dice.py:30]: changed

File: telegram/message.py (hunk 16 above)
Comment (Bibo-Joshi) [reply telegram/message.py:109]: changed

File: telegram/message.py (hunk 17 above)
Comment (Bibo-Joshi) [reply telegram/message.py:822]: Got me :D changed.

File: telegram/ext/filters.py (hunk 15 above)
Comment (Bibo-Joshi) [reply telegram/ext/filters.py:987]: While this is slightly more accurate, the sentence feels kinda weird to me. And it should be clear from the context, that only dice messages are allowed. I'll not change for now. If this is important to you, you can PR for it ;)

**Comment by Bibo-Joshi** [comment]
CI fails unrelated. merging.

**Pull request merged**